driver-core/src/main/java/com/datastax/driver/core/Frame.java

```diff
@@ -328,4 +332,16 @@ protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out)
             }
         }
     }
+
+    @ChannelHandler.Sharable
+    static class BetaVersionFlagAdder extends MessageToMessageEncoder<Frame> {
+        @Override
+        protected void encode(ChannelHandlerContext ctx, Frame frame, List<Object> out) throws Exception {
+            ProtocolVersion protocolVersion = frame.header.version;
+            // Check for the protocol version, if the downgrade has occurred during renegotiation
+            if (protocolVersion.toInt() >= ProtocolVersion.V5.toInt())
+                frame.header.flags.add(Header.Flag.USE_BETA);
```

Comment:
> Could negotiation somehow succeed and result in v5/no beta? That would be bad because the driver could fail on some of the not-yet-implemented-on-the-client v5 features.

Indeed you have a point here.

Some background first: so far we've been working on the following assumptions:

1. The driver should have `Protocol.NEWEST_SUPPORTED` pointing to the highest version available, even if it is a beta one.
1. Suppose that current driver has versions: X beta, X-1 stable
2. `Protocol.NEWEST_SUPPORTED` points to X, which is beta
3. Driver requests connection with `Protocol.NEWEST_SUPPORTED` = X (beta)

Expected Results:
<table>
<tr>
<th></th>
<th>1. Without USE_BETA flag</th>
<th>2. With USE_BETA flag</th>
</tr>
<tr>
<th>A. X is beta server-side</th>
<td>Connects with X-1 (after renegotiation) <br/> :white_check_mark:</td>
<td>Connects with X <br/> :white_check_mark:</td>
</tr>
<tr>
<th>B. X is not beta server-side</th>
<td>Connects with X <br/> :x:</td>
<td>Connects with X <br/> :warning:</td>
</tr>
</table>

As we can see, our initial assumptions had problems in some cases: if the server is newer and has X stable, our driver would most likely fail to connect if `USE_BETA` was not requested (case B1). It would probably also fail to connect even if the flag was requested (case B2), but that's more acceptable because `USE_BETA` implies "Bad Things ™️  can happen".

So I would suggest the following improvements:

1. It would indeed be better if the driver completely prevent B1 from happening at all. That would mean the driver would have to store information about beta versions:
    1. `Protocol.NEWEST_SUPPORTED` would be the last stable (V4)
    2. `Protocol.NEWEST_BETA` would be the last beta, or null
    3. If the user does not specify a version and do not call `allowBetaVersions()`, use `Protocol.NEWEST_SUPPORTED`
    4. If the user does not specify a version and _does_ call `allowBetaVersions()`, use `Protocol.NEWEST_BETA`
2. `BetaVersionFlagAdder` might become obsolete in this new design; the flag could be added in `ProtocolEncoder` and `com.datastax.driver.core.Message.Request` could convey the request for the flag, pretty much like `isTracingRequested()`.

An alternative solution would be to have the _server_ reject B1 and B2 on the grounds that from its point of view, X is not beta. That would also simplify things a lot, but I don't know if it is still time to amend the code server-side.